File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/ViewSwazzledHooks.java
package io.embrace.android.embracesdk;

import androidx.annotation.NonNull;

import io.embrace.android.embracesdk.annotation.InternalApi;
import io.embrace.android.embracesdk.internal.EmbraceInternalApi;
import io.embrace.android.embracesdk.internal.payload.TapBreadcrumb.TapBreadcrumbType;
import kotlin.Pair;

/**
 * @hide
 */
@InternalApi
public final class ViewSwazzledHooks {

    private static final String UNKNOWN_ELEMENT_NAME = "Unknown element";

    private ViewSwazzledHooks() {
    }

    static void logOnClickEvent(android.view.View view, TapBreadcrumbType breadcrumbType) {
        try {
            String viewName = "";
            try {
                viewName = view.getResources().getResourceName(view.getId());
            } catch (Exception e) {
                viewName = UNKNOWN_ELEMENT_NAME;
            }
            Pair<Float, Float> point = null;
            try {
                point = new Pair<>(view.getX(), view.getY());
            } catch (Exception e) {
                point = new Pair<>(0.0F, 0.0F);
            }
            EmbraceInternalApi.getInstance().getInternalInterface().logTap(point, viewName, breadcrumbType);
        } catch (NoSuchMethodError error) {
            // The customer may be overwriting View with their own implementation, and some of the
            // methods we use are missing.
            logError(error);
        } catch (Exception exception) {
            logError(exception);
        }
    }

    private static void logError(@NonNull Throwable throwable) {
        EmbraceInternalApi.getInstance().getInternalInterface().logInternalError(throwable);
    }

    @InternalApi
    public static final class OnClickListener {
        private OnClickListener() {
        }

        @SuppressWarnings("MethodNameCheck")
        public static void _preOnClick(android.view.View.OnClickListener thiz, android.view.View view) {
            logOnClickEvent(view, TapBreadcrumbType.TAP);
        }
    }

    @InternalApi
    public static final class OnLongClickListener {
        private OnLongClickListener() {
        }

        @SuppressWarnings("MethodNameCheck")
        public static void _preOnLongClick(android.view.View.OnLongClickListener thiz, android.view.View view) {
            if (thiz != null) {
                logOnClickEvent(view, TapBreadcrumbType.LONG_PRESS);
            }
        }
    }
}

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/ViewSwazzledHooks.kt
package io.embrace.android.embracesdk

import android.view.View
import io.embrace.android.embracesdk.annotation.InternalApi
import io.embrace.android.embracesdk.internal.EmbraceInternalApi
import io.embrace.android.embracesdk.internal.payload.TapBreadcrumb.TapBreadcrumbType

/**
 * @hide
 */
@InternalApi
public object ViewSwazzledHooks {
    private const val UNKNOWN_ELEMENT_NAME = "Unknown element"

    private fun logOnClickEvent(view: View, breadcrumbType: TapBreadcrumbType) {
        try {
            val viewName = try {
                view.resources.getResourceName(view.id)
            } catch (e: Exception) {
                UNKNOWN_ELEMENT_NAME
            }
            val point: Pair<Float?, Float?> = try {
                Pair(view.x, view.y)
            } catch (e: Exception) {
                Pair(0.0f, 0.0f)
            }
            EmbraceInternalApi.getInstance().internalInterface.logTap(
                point,
                viewName,
                breadcrumbType
            )
        } catch (error: NoSuchMethodError) {
            // The customer may be overwriting View with their own implementation, and some of the
            // methods we use are missing.
            logError(error)
        } catch (exception: Exception) {
            logError(exception)
        }
    }

    private fun logError(throwable: Throwable) {
        EmbraceInternalApi.getInstance().internalInterface.logInternalError(throwable)
    }

    @InternalApi
    public object OnClickListener {
        @Suppress("UNUSED_PARAMETER")
        @JvmStatic
        public fun _preOnClick(thiz: View.OnClickListener?, view: View) {
            logOnClickEvent(view, TapBreadcrumbType.TAP)
        }
    }

    @InternalApi
    public object OnLongClickListener {
        @JvmStatic
        public fun _preOnLongClick(thiz: View.OnLongClickListener?, view: View) {
            if (thiz != null) {
                logOnClickEvent(view, TapBreadcrumbType.LONG_PRESS)
            }
        }
    }
}
